File: backend/app/domain/user_preferences.py
from typing import Any, Literal

DateDisplayFormat = Literal["raw", "local"]
LayoutDensity = Literal["comfortable", "compact"]
ColorScheme = Literal["dark", "light", "system"]

DEFAULT_USER_PREFERENCES: dict[str, Any] = {
    "date_display_format": "raw",
    "layout_density": "comfortable",
    "sidebar_expanded": True,
    "color_scheme": "dark",
}

_VALID_DATE_DISPLAY_FORMATS = {"raw", "local"}
_VALID_LAYOUT_DENSITIES = {"comfortable", "compact"}
_VALID_COLOR_SCHEMES = {"dark", "light", "system"}


def normalize_user_preferences(stored: dict[str, Any] | None) -> dict[str, Any]:
    """Merge stored preferences with defaults and coerce to supported values."""
    merged = {**DEFAULT_USER_PREFERENCES, **(stored or {})}

    date_display_format = merged.get("date_display_format")
    if date_display_format not in _VALID_DATE_DISPLAY_FORMATS:
        date_display_format = DEFAULT_USER_PREFERENCES["date_display_format"]

    layout_density = merged.get("layout_density")
    if layout_density not in _VALID_LAYOUT_DENSITIES:
        layout_density = DEFAULT_USER_PREFERENCES["layout_density"]

    sidebar_expanded = merged.get("sidebar_expanded")
    if isinstance(sidebar_expanded, str):
        lowered = sidebar_expanded.lower()
        if lowered == "true":
            sidebar_expanded = True
        elif lowered == "false":
            sidebar_expanded = False
        else:
            sidebar_expanded = DEFAULT_USER_PREFERENCES["sidebar_expanded"]
    elif sidebar_expanded is not True and sidebar_expanded is not False:
        sidebar_expanded = DEFAULT_USER_PREFERENCES["sidebar_expanded"]

    color_scheme = merged.get("color_scheme")
    if color_scheme not in _VALID_COLOR_SCHEMES:
        color_scheme = DEFAULT_USER_PREFERENCES["color_scheme"]

    return {
        "date_display_format": date_display_format,
        "layout_density": layout_density,
        "sidebar_expanded": sidebar_expanded,
        "color_scheme": color_scheme,
    }


def merge_user_preferences_update(
    current: dict[str, Any] | None,
    patch: dict[str, Any],
) -> dict[str, Any]:
    """Apply a partial preferences update and return the normalized result."""
    base = normalize_user_preferences(current)
    update: dict[str, Any] = {}

    if "date_display_format" in patch:
        value = patch["date_display_format"]
        if value in _VALID_DATE_DISPLAY_FORMATS:
            update["date_display_format"] = value

    if "layout_density" in patch:
        value = patch["layout_density"]
        if value in _VALID_LAYOUT_DENSITIES:
            update["layout_density"] = value

    if "sidebar_expanded" in patch:
        value = patch["sidebar_expanded"]
        if isinstance(value, str):
            value = value.lower() == "true"
        if isinstance(value, bool):
            update["sidebar_expanded"] = value

    if "color_scheme" in patch:
        value = patch["color_scheme"]
        if value in _VALID_COLOR_SCHEMES:
            update["color_scheme"] = value

    return normalize_user_preferences({**base, **update})

File: frontend/src/settings/userPreferences.ts
export type DateDisplayFormat = 'raw' | 'local';
export type LayoutDensity = 'comfortable' | 'compact';
export type ColorScheme = 'dark' | 'light' | 'system';
export type ResolvedTheme = 'dark' | 'light';

export interface UserPreferences {
  dateDisplayFormat: DateDisplayFormat;
  layoutDensity: LayoutDensity;
  sidebarExpanded: boolean;
  colorScheme: ColorScheme;
}

export interface UserPreferencesApi {
  date_display_format: DateDisplayFormat;
  layout_density: LayoutDensity;
  sidebar_expanded: boolean;
  color_scheme: ColorScheme;
}

export const DEFAULT_USER_PREFERENCES: UserPreferences = {
  dateDisplayFormat: 'raw',
  layoutDensity: 'comfortable',
  sidebarExpanded: true,
  colorScheme: 'dark',
};

const LEGACY_PREFS_KEY = 'openflake.userPreferences';
const LEGACY_SIDEBAR_KEY = 'openflake.sidebar.expanded';

const VALID_COLOR_SCHEMES = new Set<ColorScheme>(['dark', 'light', 'system']);

export function getSystemTheme(): ResolvedTheme {
  if (typeof window === 'undefined') return 'dark';
  return window.matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light';
}

export function resolveTheme(colorScheme: ColorScheme): ResolvedTheme {
  if (colorScheme === 'system') return getSystemTheme();
  return colorScheme;
}

export function fromApiPreferences(
  api: Partial<UserPreferencesApi> | null | undefined,
): UserPreferences {
  return {
    dateDisplayFormat:
      api?.date_display_format === 'local' ? 'local' : DEFAULT_USER_PREFERENCES.dateDisplayFormat,
    layoutDensity:
      api?.layout_density === 'compact' ? 'compact' : DEFAULT_USER_PREFERENCES.layoutDensity,
    sidebarExpanded:
      typeof api?.sidebar_expanded === 'boolean'
        ? api.sidebar_expanded
        : DEFAULT_USER_PREFERENCES.sidebarExpanded,
    colorScheme:
      api?.color_scheme && VALID_COLOR_SCHEMES.has(api.color_scheme)
        ? api.color_scheme
        : DEFAULT_USER_PREFERENCES.colorScheme,
  };
}

export function toApiPreferences(preferences: UserPreferences): UserPreferencesApi {
  return {
    date_display_format: preferences.dateDisplayFormat,
    layout_density: preferences.layoutDensity,
    sidebar_expanded: preferences.sidebarExpanded,
    color_scheme: preferences.colorScheme,
  };
}

export function readLegacyPreferences(): Partial<UserPreferences> | null {
  const partial: Partial<UserPreferences> = {};
  let hasLegacy = false;

  try {
    const raw = localStorage.getItem(LEGACY_PREFS_KEY);
    if (raw) {
      const parsed = JSON.parse(raw) as Partial<UserPreferences>;
      if (parsed.dateDisplayFormat === 'local' || parsed.dateDisplayFormat === 'raw') {
        partial.dateDisplayFormat = parsed.dateDisplayFormat;
        hasLegacy = true;
      }
      if (parsed.layoutDensity === 'compact' || parsed.layoutDensity === 'comfortable') {
        partial.layoutDensity = parsed.layoutDensity;
        hasLegacy = true;
      }
    }

    const sidebar = localStorage.getItem(LEGACY_SIDEBAR_KEY);
    if (sidebar === 'true' || sidebar === 'false') {
      partial.sidebarExpanded = sidebar === 'true';
      hasLegacy = true;
    }
  } catch {
    return null;
  }

  return hasLegacy ? partial : null;
}

export function clearLegacyPreferences(): void {
  localStorage.removeItem(LEGACY_PREFS_KEY);
  localStorage.removeItem(LEGACY_SIDEBAR_KEY);
}

export function applyLayoutDensity(density: LayoutDensity): void {
  document.documentElement.dataset.layoutDensity = density;
}

export function applyColorScheme(colorScheme: ColorScheme): () => void {
  const root = document.documentElement;

  const apply = () => {
    const resolved = resolveTheme(colorScheme);
    root.dataset.theme = resolved;
    root.dataset.colorScheme = colorScheme;
  };

  apply();

  if (colorScheme !== 'system') {
    return () => {};
  }

  const media = window.matchMedia('(prefers-color-scheme: dark)');
  const onChange = () => apply();
  media.addEventListener('change', onChange);
  return () => media.removeEventListener('change', onChange);
}
